src/overlay/SurveyDataManager.cpp

```diff
@@ -0,0 +1,484 @@
+                                  TopologyResponseBodyV2& response)
+{
+    ZoneScoped;
+    std::lock_guard<std::recursive_mutex> const guard(mMutex);
+
+    if (phase() == SurveyPhase::REPORTING && mNonce == request.nonce &&
+        mSurveyor == request.request.surveyorPeerID)
+    {
+        if (!mFinalNodeData.has_value())
+        {
+            emitInconsistencyError("getSurveyData");
+            return false;
+        }
+
+        response.nodeData = mFinalNodeData.value();
+        response.inboundPeers = fillTimeSlicedPeerDataList(
+            mFinalInboundPeerData,
+            static_cast<size_t>(request.inboundPeersIndex));
+        response.outboundPeers = fillTimeSlicedPeerDataList(
+            mFinalOutboundPeerData,
+            static_cast<size_t>(request.outboundPeersIndex));
+        return true;
+    }
+    return false;
+}
+
+bool
+SurveyDataManager::surveyIsActive()
+{
+    std::lock_guard<std::recursive_mutex> const guard(mMutex);
+    return phase() != SurveyPhase::INACTIVE;
+}
+
+#ifdef BUILD_TESTS
+void
+SurveyDataManager::setPhaseMaxDurationsForTesting(
+    std::chrono::minutes maxPhaseDuration)
+{
+    std::lock_guard<std::recursive_mutex> const guard(mMutex);
+    mMaxPhaseDurationForTesting = maxPhaseDuration;
+}
+#endif
+
+SurveyPhase
+SurveyDataManager::phase()
+{
+    if (mRecordStartTime.has_value() && !mRecordEndTime.has_value())
+    {
+        if (mClock.now() >
+            mRecordStartTime.value() + getRecordingPhaseMaxDuration())
+        {
+            CLOG_INFO(
+                Overlay,
+                "Survey recording phase has expired. Resetting survey data.");
+            reset();
+            return SurveyPhase::INACTIVE;
+        }
+        return SurveyPhase::RECORDING;
+    }
+    if (mRecordStartTime.has_value() && mRecordEndTime.has_value())
+    {
+        if (mClock.now() >
+            mRecordEndTime.value() + getReportingPhaseMaxDuration())
+        {
+            CLOG_INFO(
+                Overlay,
+                "Survey reporting phase has expired. Resetting survey data.");
+            reset();
+            return SurveyPhase::INACTIVE;
+        }
+        return SurveyPhase::REPORTING;
+    }
+    return SurveyPhase::INACTIVE;
+}
+
+void
+SurveyDataManager::reset()
+{
+    mRecordStartTime.reset();
+    mRecordEndTime.reset();
+    mNonce.reset();
+    mSurveyor.reset();
+    mRecordingNodeData.reset();
+    mRecordingInboundPeerData.clear();
+    mRecordingOutboundPeerData.clear();
+    mFinalNodeData.reset();
+    mFinalInboundPeerData.clear();
+    mFinalOutboundPeerData.clear();
+}
+
+void
+SurveyDataManager::emitInconsistencyError(std::string const& where)
+{
+    CLOG_ERROR(Overlay, "Encountered inconsistent survey data while executing "
+                        "`{}`. Resetting survey state.");
+    reset();
+}
+
+UnorderedMap<NodeID, RecordingPeerData>&
+SurveyDataManager::getPeerDataMapFromPeer(Peer const& peer)
+{
+    switch (peer.getRole())
```

Comment:
I'd recommend SurveyManager passing inbound and outbound peers to SurveyDataManager separately, so we don't have to query Peer's internals. I know it's in the public API, but this will likely to change in the future as we're making parallelization changes that necessitate stronger encapsulation. You can use `OverlayManager::getInboundAuthenticatedPeers` and `OverlayManager::getOutboundAuthenticatedPeers` for this.

Reply:
Done in 63666f792a1fc2d6f1b2f7de3bfd83acb0597b7d

Reply:
Ah, that makes sense. Fixed in ce0efda152dcb309cecf908416035413c29f4ba5